Added support for signer

exp-table opened the pull request.

fracek commented on src/nile/commands/proxy.py:
```diff
@@ -0,0 +1,68 @@
+        signer_data = next(accounts.load(str(signer.public_key), network))
+        signer.account = signer_data["address"]
+        signer.index = signer_data["index"]
+    else: #doesn't exist, havbe to deploy
```
There should be a space between `#` and `doesn't`. There's a type (`havbe` should be `have`).

fracek commented on src/nile/commands/proxy.py:
```diff
@@ -0,0 +1,68 @@
+        signer.index = accounts.current_index(network)
+        subprocess.run(f"nile deploy Account {signer.public_key} --alias account-{signer.index}", shell=True)
+        address, _ = next(deployments.load(f"account-{signer.index}", network))
+        #initialize account
```
Space between `#` and `initialize`

fracek commented:
I like this! I added a couple of nitpicky comments.

exp-table commented:
> I like this! I added a couple of nitpicky comments.

Fixed it!

fracek commented:
Ok there are a couple of linting issues. My advice is to run `tox -e format` to automatically format everything, then `tox -e lint` to double check everything is formatted correctly.

martriay commented:
This looks very nice! Do you think `proxy` could be merged into `send`? Or maybe I'm not understanding the different use case. In any case, I don't think that's the right name for it since it refers to a different kind of contracts which delegate execution to other contracts. Maybe `raw_execute`?

Future work could be:

1. auto generating a few accounts on `nile node`, have a default one and `--account` to select by index or alias
2. generating a private key, needed for `1` and pretty useful

exp-table commented on src/nile/commands/proxy.py:
```diff
@@ -0,0 +1,68 @@
+        signer.index = signer_data["index"]
+    else: # doesn't exist, have to deploy
+        signer.index = accounts.current_index(network)
+        subprocess.run(f"nile deploy Account {signer.public_key} --alias account-{signer.index}", shell=True)
```
I didn't even think about it when prototyping initially. Good catch, fixed it!

exp-table commented:
> This looks very nice! A few comments:
> 
>     * Do you think `proxy` could be merged into `send`? Or maybe I'm not understanding the different use case. In any case, I don't think that's the right name for it since it refers to a different kind of contracts which delegate execution to other contracts. Maybe `raw_execute`?
> 
>     * Could you expand the documentation to include an example env file and commands snippets with actual parameters? That help users a lot.
> 

I renamed them to `send` and `raw-execute`. As you said, `proxy` isn't the right label for what it does.
I also added an example `example.env` file as well as enhanced documentations for `send` and `raw-execute`.

fracek reviewed (approved):
Looks good to me. All issues have been addressed, thanks!

fracek commented on src/nile/signer.py:
```diff
@@ -0,0 +1,48 @@
+"""Utility for sending signed transactions to an Account on Starknet."""
+import subprocess
+
+from starkware.cairo.common.hash_state import compute_hash_on_elements
```
This should be be wrapped in a try catch to avoid throwing when users don't have cairo installed yet (and want to use nile to install it).

fracek commented:
One last thing before I'm ready to merge. Sorry for letting you wait so long to review this PR!

martriay commented on README.md:
```diff
@@ -88,6 +88,56 @@ A few things to notice here:
+
+A few things to notice here:
+
+1. `nile set <env_var>` looks for an environement variable with the same name whose value is a private key
```
there's no `nile set` command :)

fracek commented on README.md:
```diff
@@ -88,6 +88,56 @@ A few things to notice here:
 4. By default Nile works on local, but you can pass `--network mainnet` to deploy directly to a public chain! Notice that `mainnet` refers to StarkNet main chain, that's settled on Goerli testnet of Ethereum ([mainnet deployment this month!](https://medium.com/starkware/starknet-alpha-is-coming-to-mainnet-b825829eaf32))
 
+### `setup`
+You can find an exemple `.env` file in `example.env`. These are private keys only to be used for testing and never in production.
```
`exemple` -> `example`

fracek commented on src/nile/signer.py:
```diff
@@ -0,0 +1,55 @@
+    def get_nonce(self):
+        """Get the nonce for the next transaction."""
+        nonce = subprocess.check_output(
+            f"nile call account-{self.index} get_nonce", shell=True, encoding="utf-8"
```
Here it's missing a `--network` argument so when using a non-default network this call fails

fracek reviewed (changes requested):
Looks like it works well on a local network but does not work on mainnet because the network argument is missing in the call to get the nonce.

fracek reviewed (approved):
I run this in a fresh venv and everything is working as expected

martriay commented:
Thanks all 🎉
